File: exec/utils/runguard.py
import sys, argparse, os

class Range(object):
    def __init__(self, s):
        super.__init__(self)

        splitted = s.split(':')
        left = 0
        right = 0
        if len(splitted) >= 1:
            left = int(splitted[0])
        if len(splitted) >= 2:
            right = int(splitted[1])
        
        if left > right or left < 0 or right < 0:
            raise ValueError('must be in form <soft:positive number>:<hard:positive number>')
        
        self.soft = left
        self.hard = right

def main():
    parser = argparse.ArgumentParser(prog='runguard', description='Converts command line arguments into runc config.json')
    parser.add_argument('-r', '--rootless', action='store_true', help='start container ')
    parser.add_argument('-r', '--root', action='store', type=str, help='rootfs', required=True)
    parser.add_argument('-u', '--user', action='store', type=int, help='specifies user id', required=True)
    parser.add_argument('-g', '--group', action='store', type=int, help='specifies group id', required=True)
    parser.add_argument('--netns', action='store', type=str, help='specifies network namespace file name to inherit namespace from')
    parser.add_argument('-d', '--work-dir', action='store', type=str, help='run command in specified network namespace, if not specified, runguard will unshare network namespace by default')
    parser.add_argument('-T', '--wall-time', action='store', type=Range, help='kill command after wall time clock seconds (floating point is accpeted)')
    parser.add_argument('-t', '--cpu-time', action='store', type=int, help='kill command after CPU time clock seconds (floating point is accepted)', default='0')
    parser.add_argument('-m', '--memory-limit', action='store', type=int, help='maximum kilo-bytes that all processes in container may consume')
    parser.add_argument('-f', '--file-limit', action='store', type=int, help='maximum kilo-bytes that all processes in container can write', default='0')
    parser.add_argument('-p', '--nproc', action='store', type=int, help='maximum number of processes may run in container', default='0')
    parser.add_argument('-P', '--cpuset', action='store', type=str, help='processor IDs that processes in container may run on')
    parser.add_argument('--allowed-syscall', action='store', type=str, help='whitelisted syscalls that processes in container may invoke')
    parser.add_argument('-c', '--no-core-dumps', action='store_true', help='disable core dumps in container')
    parser.add_argument('-E', '--environment', action='store_true', help='pass through exisitng environment variables into container (exluding PATH)')
    parser.add_argument('-V', '--variable', action='append', type=str, help='set additional environment variable (eg. -Vkey1=value1 -Vkey2=value2)')
    parser.add_argument('commands', metavar='N', type=str, nargs='+', help='command line of program to run in container')
    args = parser.parse_args()

    env = []
    if args.environment:
        env += os.environ
    if args.variable:
        print(args.variable)
    env = [x for x in env if not x.startsWith('PATH=')]
    env += ['PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin']

    config = {
        'ociVersion': '1.0.1-dev',
        'process': {
            'terminal': False,
            'args': args.commands,
            'env': env,
            'cwd': args.work_dir or '/',
            'capabilities': [

            ],
            'rlimits': ([
                {
                    'type': 'RLIMIT_AS',
                    'hard': args.memory_limit * 1024,
                    'soft': args.memory_limit * 1024,
                },
                {
                    'type': 'RLIMIT_DATA',
                    'hard': args.memory_limit * 1024,
                    'soft': args.memory_limit * 1024,
                }
            ] if args.memory_limit > 0 else []) + ([
                {
                    'type': 'RLIMIT_FSIZE',
                    'hard': args.file_limit,
                    'soft': args.file_limit,
                }
            ] if args.file_limit > 0 else []) + ([
                {
                    'type': 'RLIMIT_CPU',
                    'hard': args.cpu_limit,
                    'soft': args.cpu_limit,
                }
            ] if args.cpu_limit > 0 else []) + ([
                {
                    'type': 'RLIMIT_NPROC',
                    'hard': args.nproc,
                    'soft': args.nproc
                }
            ] if args.nproc > 0 else []) + ([
                {
                    'type': 'RLIMIT_CORE',
                    'hard': 0,
                    'soft': 0
                }
            ] if args.no_core_dumps else []),
            'noNewPrivileges': True
        },
        'root': {
            'path': args.root,
            'readonly': True
        },
        'hostname': 'judge',
        'mounts': [
            {
                'destination': '/proc',
                'type': 'proc',
                'source': 'proc'
            },
            {
                'destination': '/dev',
                'type': 'tmpfs',
                'source': 'tmpfs',
                'options': [
                    'nosuid',
                    'strictatime',
                    'mode=755',
                    'size=65536k'
                ]
            },
            {
                'destination': '/dev/pts',
                'type': 'devpts',
                'source': 'devpts',
                'options': [
                    'nosuid',
                    'noexec',
                    'newinstance',
                    'ptmxmode=0666',
                    'mode=0620'
                ]
            },
            {
                'destination': '/dev/shm',
                'type': 'tmpfs',
                'source': 'shm',
                'options': [
                    'nosuid',
                    'noexec',
                    'nodev',
                    'mode=1777',
                    'size=65536k'
                ]
            },
            {
                'destination': '/dev/mqueue',
                'type': 'mqueue',
                'source': 'mqueue',
                'options': [
                    'nosuid',
                    'noexec',
                    'nodev'
                ]
            },
            {
                'destination': '/sys',
                'type': 'none',
                'source': '/sys',
                'options': [
                    'rbind',
                    'nosuid',
                    'noexec',
                    'nodev',
                    'ro'
                ]
            }
        ],
        'linux': {
            "uidMappings": [
                {
                    "containerID": 1000,
                    "hostID": args.user,
                    "size": 1
                }
            ],
            "gidMappings": [
                {
                    "containerID": 1000,
                    "hostID": args.group,
                    "size": 1
                }
            ],
            "namespaces": [
                {
                    "type": "pid"
                },
                {
                    "type": "ipc"
                },
                {
                    "type": "uts"
                },
                {
                    "type": "mount"
                },
                {
                    "type": "user"
                }
            ],
        }
    }


if __name__ == '__main__':
    main()

File: runguard/src/runguard.cpp
#include <boost/lexical_cast.hpp>
#include <boost/log/attributes.hpp>
#include <boost/log/common.hpp>
#include <boost/log/core.hpp>
#include <boost/log/expressions.hpp>
#include <boost/log/sinks.hpp>
#include <boost/log/support/date_time.hpp>
#include <boost/log/trivial.hpp>
#include <boost/log/utility/setup/common_attributes.hpp>
#include <boost/log/utility/setup/console.hpp>
#include <boost/log/utility/setup/file.hpp>
#include <boost/program_options.hpp>
#include <boost/stacktrace.hpp>
#include <filesystem>
#include <fstream>
#include <iostream>

#include "run.hpp"
#include "system.hpp"
#include "utils.hpp"

using namespace std;

// for debug
std::ostream& operator<<(std::ostream& out, runguard_options& opt) {
    out << "croupname = " << opt.cgroupname << " chroot_dir = " << opt.chroot_dir << " work_dir = " << opt.work_dir << " preexecute = " << opt.preexecute << " user = " << opt.user << " group = " << opt.group << " cpuset = " << opt.cpuset << " metafile_path = " << opt.metafile_path;
    return out;
}

void validate(boost::any& v, const vector<string>& values, size_t*, int) {
    using namespace boost::program_options;
    validators::check_first_occurrence(v);

    string const& s = validators::get_single_string(values);
    if (s[0] == '-') {
        throw validation_error(validation_error::invalid_option_value);
    }

    v = boost::lexical_cast<size_t>(s);
}

void validate(boost::any& v, const vector<string>& values, struct time_limit*, int) {
    using namespace boost::program_options;
    validators::check_first_occurrence(v);

    struct time_limit result;
    string const& s = validators::get_single_string(values);
    auto colon = s.find(':');
    string left = s.substr(0, colon);
    string right = colon < s.size() ? s.substr(colon + 1) : "";

    result.soft = boost::lexical_cast<double>(left);
    if (right.size())
        result.hard = boost::lexical_cast<double>(right);
    else
        result.hard = result.soft;

    if (result.hard < result.soft ||
        !finite(result.hard) || !finite(result.soft) ||
        result.hard < 0 || result.soft < 0)
        throw validation_error(validation_error::invalid_option_value);

    v = result;
}

void init_boost_log() {
    boost::log::add_common_attributes();
    auto core = boost::log::core::get();

    core->add_global_attribute("UTCTimeStamp", boost::log::attributes::utc_clock());

    // clang-format off
    auto log_format(
        boost::log::expressions::stream << 
            "[" << boost::log::expressions::format_date_time<boost::posix_time::ptime>("TimeStamp", "%Y-%m-%d %H:%M:%S") << 
            "] [" << boost::log::expressions::attr<boost::log::attributes::current_thread_id::value_type>("ThreadID") << 
            "] [" << std::left << std::setw(7) << std::setfill(' ') << boost::log::trivial::severity << 
            "] " << boost::log::expressions::smessage 
        );
    
    boost::log::add_file_log(
        boost::log::keywords::file_name = "runguard_%d_%m_%Y.%N.log",
        boost::log::keywords::rotation_size = 1 * 1024 * 1024,  // 每1M滚动一次日志
        boost::log::keywords::target = filesystem::path(getenv("BOOST_log_dir")) / "runguard",
        boost::log::keywords::min_free_space = 30 * 1024 * 1024, // 磁盘至少有30M空间
        boost::log::keywords::max_size = 20 * 1024 * 1024, // 最多存储20M日志
        boost::log::keywords::time_based_rotation = boost::log::sinks::file::rotation_at_time_point(boost::gregorian::greg_day(1)), // 或者每月1号滚动日志
        boost::log::keywords::scan_method = boost::log::sinks::file::scan_matching,
        boost::log::keywords::format = log_format,  // "[%TimeStamp%] [%Severity%]: %Message%",
        boost::log::keywords::auto_flush = true);
    // clang-format on

    boost::log::add_console_log(std::cout, boost::log::keywords::format = log_format);
}

int main(int argc, const char* argv[]) {
    ofstream out("/var/log/judge-system/runguard/label", std::ofstream::out);
    string s = "BOOST_log_dir = " + string{filesystem::path(getenv("BOOST_log_dir")).u8string()};
    out.write(s.c_str(), 100);

    init_boost_log();

    namespace po = boost::program_options;
    po::options_description desc("runguard options");
    po::positional_options_description pos;
    po::variables_map vm;

    struct runguard_options opt;

    // clang-format off
    desc.add_options()
        ("root,r", po::value<string>(), "run command with root directory set to root. If this option is provided, running command is executed relative to the chroot.")
        ("user,u", po::value<string>(), "run command as user with username or user id")
        ("work", po::value<string>(), "work directory for command")
        ("group,g", po::value<string>(), "run command under group with groupname or group id. If only 'user' is set, this defaults to the same")
        ("netns", po::value<string>(), "run command in specified network namespace, if not specified, runguard will create a new network namespace every time")
        ("wall-time,T", po::value<time_limit>(), "kill command after wall time clock seconds (floating point is acceptable)")
        ("cpu-time,t", po::value<time_limit>(), "set maximum CPU time (floating point is acceptable) consumption of the command in seconds")
        ("memory-limit,m", po::value<size_t>(), "set maximum memory consumption of the command in KB")
        ("file-limit,f", po::value<size_t>(), "set maximum created file size of the command in KB")
        ("nproc,p", po::value<size_t>(), "set maximum process living simutanously")
        ("cpuset,P", po::value<string>(), "set the processor IDs that can only be used (e.g. \"0,2-3\")")
        ("allowed-syscall", po::value<string>(), "set the limited syscall numbers in file separated by spaces")
        ("no-core-dumps,c", "disable core dumps")
        ("preexecute", po::value<string>(), "run command in new mount namespace before user program execution")
        ("standard-input-file,i", po::value<string>(), "redirect command standard input fd to file")
        ("standard-output-file,o", po::value<string>(), "redirect command standard output fd to file")
        ("standard-error-file,e", po::value<string>(), "redirect command standard error fd to file")
        ("environment,E", "preseve system environment variables (or only PATH is loaded)")
        ("variable,V", po::value<vector<string>>(), "add additional environment variables (e.g. -Vkey1=value1 -Vkey2=value2)")
        ("out-meta,M", po::value<string>(), "write runguard monitor results (run time, exitcode, memory usage, ...) to file")
        ("cmd", po::value<vector<string>>()->composing()->required(), "commands")
        ("help", "display this help text")
        ("version", "display version of this application");
    // clang-format on

    pos.add("cmd", -1);

    try {
        po::store(po::command_line_parser(argc, argv)
                      .options(desc)
                      .positional(pos)
                      .allow_unregistered()
                      .run(),
                  vm);
        po::notify(vm);
    } catch (po::error& e) {
        cerr << e.what() << endl
             << endl;
        cerr << desc << endl;
        return 1;
    }

    if (vm.count("help")) {
        cout << "Runguard: Running user program in protected mode with system resource access limitations." << endl
             << "This app requires root privilege if either 'root' or 'user' option is provided." << endl
             << "Usage: " << argv[0] << " [options] -- [command]";
        cout << desc << endl;
        return 0;
    }

    if (vm.count("version")) {
        cout << "runguard" << endl;
        return 0;
    }

    if (vm.count("root")) {
        opt.chroot_dir = vm["root"].as<string>();
    }

    if (vm.count("user")) {
        opt.user = vm["user"].as<string>();
        opt.user_id = get_userid(opt.user.c_str());
        opt.group_id = get_groupid(opt.user.c_str());  // debug: default the same
    }

    // if (vm.count("group") || vm.count("user")) { // origin
    if (vm.count("group")) {  // debug
        opt.group = vm["group"].as<string>();
        opt.group_id = get_groupid(opt.group.c_str());
    }

    if (vm.count("netns")) opt.netns = vm["netns"].as<string>();
    if (vm.count("preexecute")) opt.preexecute = vm["preexecute"].as<string>();
    if (vm.count("work")) opt.work_dir = vm["work"].as<string>();

    if (vm.count("variable")) {
        opt.env = vm["variable"].as<vector<string>>();
    }

    if (vm.count("wall-time")) opt.use_wall_limit = true, opt.wall_limit = vm["wall-time"].as<time_limit>();
    if (vm.count("cpu-time")) {
        opt.use_cpu_limit = true, opt.cpu_limit = vm["cpu-time"].as<time_limit>();
        if (!opt.use_wall_limit) {
            // wall limit must be taken
            opt.use_wall_limit = true;
            opt.wall_limit = opt.cpu_limit;
            opt.wall_limit.hard *= 3;
            opt.wall_limit.soft *= 3;
        }
    }
    if (vm.count("memory-limit")) {
        opt.memory_limit = vm["memory-limit"].as<size_t>();
        if (opt.memory_limit != (opt.memory_limit * 1024) / 1024)
            opt.memory_limit = -1;
        else
            opt.memory_limit *= 1024;
    }
    if (vm.count("file-limit")) {
        opt.file_limit = vm["file-limit"].as<size_t>();
        if (opt.file_limit != (opt.file_limit * 1024) / 1024)
            opt.file_limit = -1;
        else
            opt.file_limit *= 1024;
    }
    if (vm.count("nproc")) opt.nproc = vm["nproc"].as<size_t>();
    if (vm.count("no-core-dumps")) opt.no_core_dumps = true;
    if (vm.count("allowed-syscall")) {
        filesystem::path p = vm["allowed-syscall"].as<string>();
        if (filesystem::exists(p)) {
            ifstream fin(p);
            for (int no; fin >> no;) opt.syscalls.push_back(no);
        }
    }
    if (vm.count("cpuset")) opt.cpuset = vm["cpuset"].as<string>();
    if (vm.count("standard-input-file")) opt.stdin_filename = vm["standard-input-file"].as<string>();
    if (vm.count("standard-output-file")) opt.stdout_filename = vm["standard-output-file"].as<string>();
    if (vm.count("standard-error-file")) opt.stderr_filename = vm["standard-error-file"].as<string>();
    if (vm.count("environment")) opt.preserve_sys_env = true;
    if (vm.count("out-meta")) opt.metafile_path = vm["out-meta"].as<string>();
    opt.command = vm["cmd"].as<vector<string>>();

    BOOST_LOG_TRIVIAL(debug) << "opt: " << opt;

    return runit(opt);
}